### PR TITLE
Adding middleware component for non interactive oauth2 client credentials flow

### PR DESCRIPTION
**Pull request by h2floh**
# Description

Added a http middleware component, with which it is possible to retrieve a token with an oauth2 client credential flow.
In addition to existing middleware.http.oauth2 this will allow non interactive token retrieval for an application/service principal. 

This is not meant to be used on an public endpoint but within service to service invocation if the invoked service needs authorization tokens for talking with external APIs.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation : https://github.com/dapr/docs/pull/741
* [X] Added a community sample : https://github.com/dapr/samples/pull/5
* [X] Opened Draft PR for loading the component: https://github.com/dapr/dapr/pull/1899

### CODE REVIEW

**Comment by yaron2** [comment]
Thanks for this!

We also need a change [here](https://github.com/dapr/docs/blob/9d1919ef9485e140cfcdeebd676571e9a2cc0c0e/howto/authorization-with-oauth/README.md) too, right?

**Comment by yaron2** [comment]
@h2floh can you make sure you have Go 1.14+ installed, and then run `go mod tidy` and push the changes? Thanks!

**Comment by h2floh** [comment]
> Thanks for this!
> 
> We also need a change [here](https://github.com/dapr/docs/blob/9d1919ef9485e140cfcdeebd676571e9a2cc0c0e/howto/authorization-with-oauth/README.md) too, right?

Oh, yeah thanks for the hint! Will change the documentation and open a PR for it too. 

I obviously focused too much on the quickstart :)

**Comment by h2floh** [comment]
> @h2floh can you make sure you have Go 1.14+ installed, and then run `go mod tidy` and push the changes? Thanks!

@yaron2 my go version is `go1.14.7 linux/amd64`, just pushed the changes from go mod tidy.

File: middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	Scopes              string `json:"scopes"`
+	TokenURL            string `json:"tokenURL"`
+	HeaderName          string `json:"headerName"`
+	EndpointParamsQuery string `json:"endpointParamsQuery,omitempty"`
+	AuthStyleString     string `json:"authStyle"`
+	AuthStyle           int    `json:"-"`
+}
+
+// dapr logger
+var log = logger.NewLogger("dapr.components-contrib.middleware.http.oauth2clientcredentials")
```
Comment (h2floh) [inline middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go:39]: @yaron2 , I have a question regarding the dapr logger, linter is complaining about the global variable here
```
middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go:39:5  gochecknoglobals  `log` is a global variable
```
So I tried to put the logger into the `Middleware` struct, but then there was no output in daprd?! Is that a known behavior of the logging mechanism or am I doing something wrong?

Comment (yaron2) [reply middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go:39]: It needs to be a struct member, you can look at other components. A logger is already injected from daprd

Comment (yaron2) [reply middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go:39]: https://github.com/dapr/components-contrib/blob/master/middleware/http/ratelimit/ratelimit_middleware.go

Comment (h2floh) [reply middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go:39]: Thanks, missed that I can pass the logger from daprd main.go!

Comment (h2floh) [reply middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go]: Changed, tested and pushed.

**Comment by h2floh** [comment]
> We also need a change [here](https://github.com/dapr/docs/blob/9d1919ef9485e140cfcdeebd676571e9a2cc0c0e/howto/authorization-with-oauth/README.md) too, right?

@yaron2 done https://github.com/dapr/docs/pull/741

File: middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go (hunk 1 above)
Comment (h2floh) [reply middleware/http/oauth2clientcredentials/oauth2clientcredentials_middleware.go]: The only drawback I see with that approach is that you can not identify which component logged a message or error. 

Scope will always be 'dapr.runtime' 
```
== DAPR == time="2020-08-13T04:53:00.2655148Z" level=debug msg="Cached token not found, try get one" app_id=msgraphapp instance=771becc56b7d scope=dapr.runtime type=log ver=edge
```

**Pull request merged**